### ISSUE TITLE
Extracting IAM short service names for policies

### ISSUE THREAD

**Issue by lvh** [aws/aws-sdk-java#1066]
IAM uses short service names in policies, e.g. the `swf` in:

```json
{
 "Statement": [{
                "Action": ["swf:Res*"],
                "Resource": "*",
                "Effect": "Allow"
                }]
}
```

I am trying to figure out how to get to this name from Java SDK objects, e.g. client objects.

I believe this should be the `signingPrefix` key under `metadata` in the model for a particular service. It is definitely not the endpoint prefix name in the API metadata or the part before the dash in the model's file name. (Counterexample for both: AWS AppStream has IAM prefix `appstream` but endpoint prefix and model file name `appstream2`, but there are plenty more, including e.g. the IoT data API.)

There appears to be at least one counterexample to it being under `metadata signingRule` outside of the trivial S3 one (since S3 doesn't have model files): IoT. For some reason [the IoT metadata intermediate JSON](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-models/src/main/resources/models/iot-2015-05-28-intermediate.json) lists the  signing name as `execute-api` which is clearly pretty different from the other values. Additionally, iotdata APIs are listed in under the `iot:` prefix in [the IAM documentation for IoT](http://docs.aws.amazon.com/IAM/latest/UserGuide/list_iot.html), all marked with the string "- this is an IAM policy permission only, not an API action that can be called.". I'm not sure what they mean by that -- they appear to be callable [in the Java SDK](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-iot/src/main/java/com/amazonaws/services/iotdata/AWSIotDataClient.java). Meanwhile the signing name for iotdata is `iotdata`, its prefix is `data.iot` but nothing seems to use either for IAM roles.

I would've filed a bug via the console with IoT, but unlike previous issues I filed, it's not clear that this is actually a bug in their service document (i.e signing-name is supposed to be the IAM prefix too) or just a mistaken assumption about how this works. It is also not clear that it's just an IoT bug. If it turns out that this is considered a bug, I'll do more research to find all places it exists.

Given, say, a client and a method (operation name), how do I find the IAM name for that operation/method? (I could try to get this data from IAM directly, but I'd prefer to use the Java SDK as the single source of truth if at all possible. It's also not clear that IAM publishes this information in a programmatically accessible manner, but I'd love to be wrong about that. There appears to be a JSON file that [policygen](https://awspolicygen.s3.amazonaws.com/policygen.html) has, but that hardly seems like a public API.)

**Comment by dagnir** [comment]
Hi @lvh, as you've discovered there's no systematic way to determine from the model what the correct set of policy actions are (including the prefix). Our code generator does provide the option to generate an auth policy actions enum for the client: https://github.com/aws/aws-sdk-java/blob/8ca2018dbc4a7958c3626171e28ec467bd4a32a7/aws-java-sdk-code-generator/src/main/java/com/amazonaws/codegen/emitters/CodeEmitter.java#L462

However, it simply uses the set of operations as the actions, which is often wrong. This is why we skip it all together for a lot of the services e.g. [Lex](https://github.com/aws/aws-sdk-java/blob/33d67a868f0ea30fb4906b43c6515345b0f7918f/aws-java-sdk-models/src/main/resources/models/runtime.lex-2016-11-28-intermediate.json#L7). For others, like KMS, the enum is handwritten.

As far as the action prefix, by default it simply uses the endpoint prefix:

https://github.com/aws/aws-sdk-java/blob/8ca2018dbc4a7958c3626171e28ec467bd4a32a7/aws-java-sdk-code-generator/src/main/java/com/amazonaws/codegen/emitters/CodeEmitter.java#L584-L590

but as this can also be different (as you discovered), we also have an option to override it in the codegen configuration: [`actionPrefix`](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-models/src/main/resources/models/runtime.lex-2016-11-28-intermediate.json#L5)

In short, I don't think this is possible to do in general for all services.

CC: @shorea

**Comment by lvh** [comment]
OK. Are you interested in bug reports where either of those heuristics (operations for IAM actions, endpoint prefix for IAM prefix) is wrong?

**Comment by dagnir** [comment]
Sure, if you have the data on hand I it might be interesting to see. I suspect there are quite a few that are wrong, so a single issue to consolidate them would probably be good :). Going ahead and closing this issue.
